hconfig_utils/mapl3g_hconfig_get_private.F90

```diff
@@ -0,0 +1,149 @@
+      integer(kind=ESMF_KIND_I8), optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'I8'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_I8
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_i8
+
+   subroutine get_value_r4(params, value, default, valuestring, rc)
+      real(kind=ESMF_KIND_R4), intent(inout) :: value
+      real(kind=ESMF_KIND_R4), optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'R4'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_R4
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_r4
+
+   subroutine get_value_r8(params, value, default, valuestring, rc)
+      real(kind=ESMF_KIND_R8), intent(inout) :: value
+      real(kind=ESMF_KIND_R8), optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'R8'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_R8
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_r8
+
+   subroutine get_value_string(params, value, default, valuestring, rc)
+      character(len=:), allocatable, intent(inout) :: value
+      character(len=*), optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'CH'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_CH
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_string
+
+   subroutine get_value_logical(params, value, default, valuestring, rc)
+      logical, intent(inout) :: value
+      logical, optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'L'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_L
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_logical
+
+!============== Scalar subroutines must appear above this line. ================
+
+!============================= ARRAY VALUE TYPES ===============================
+#define ISARRAY 1
+!=============== Array subroutines must appear below this line. ================
+
+   subroutine get_value_i4seq(params, value, default, valuestring, rc)
+      integer(kind=ESMF_KIND_I4), dimension(:), allocatable, intent(inout) :: value
+      integer(kind=ESMF_KIND_I4), dimension(:), optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'I4'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_I4
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_i4seq
+
+   subroutine get_value_i8seq(params, value, default, valuestring, rc)
+      integer(kind=ESMF_KIND_I8), dimension(:), allocatable, intent(inout) :: value
+      integer(kind=ESMF_KIND_I8), dimension(:), optional, intent(in) :: default
+      character(len=*), parameter :: typestring = 'I8'
+      character(len=*), parameter :: edit_descriptor = EDIT_DESC_I8
+#include "mapl3g_hconfig_get_value_template.h"
+   end subroutine get_value_i8seq
```

Comment:
This approach is generally harder to understand than one where the template also includes the procedure interface.   It just means you have to also define the proc name as a token.

Reply:
True. The value and default arguments cannot appear in the template. Actually, the types, dimensions, and allocatable status cannot appear in the template. It's possible to replace these with macros:

- type of value and default
- kind of value and default
- len for value (character)
- len for default (character)
- dimension(:) for arrays
- allocatable for arrays

The downside is that the macros change for each subroutine. It seemed a bad idea to redefine the macros multiple times in the same source file. I noticed that the preprocessor reports each time they are redefined. By carefully grouping the subroutines, some of the redefinitions can be avoided, but there are still multiple redefinitions.

typestring and edit_descriptor can be set by macros that change for each subroutine

Reply:
I'll leave this for the refactor.